### PR TITLE
Combine multiple shard query requests to the same QueryNode

### PR DESCRIPTION
**Pull request by Zach41**
Merging multiple shard query/search requests into one rpc request if QueryNode is the leader of multiple shards of Collection
issue:#16823

Signed-off-by: Zach41 <zongmei.zhang@zilliz.com>

### CODE REVIEW

**Comment by Zach41** [comment]
return ut

**Comment by Zach41** [comment]
rerun ut

**Comment by Zach41** [comment]
rerun ut

File: internal/querynode/impl.go [hunk 1]
```diff
@@ -433,6 +436,56 @@ func (node *QueryNode) isHealthy() bool {
...
+		Status: &commonpb.Status{
+			ErrorCode: commonpb.ErrorCode_Success,
+		},
+	}
+	toReduceResults := make([]*internalpb.SearchResults, 0)
+	runningGp, runningCtx := errgroup.WithContext(ctx)
+	mu := &sync.Mutex{}
+	for _, ch := range req.GetDmlChannels() {
+		ch := ch
+		req := proto.Clone(req).(*querypb.SearchRequest)
```
Comment (czs007) [inline internal/querynode/impl.go:456]: Is it possible not to clone?

Comment (czs007) [reply internal/querynode/impl.go:456]: use shallow copy intead?

Comment (Zach41) [reply internal/querynode/impl.go]: > use shallow copy intead?
changed to use shallow copy

**Comment by czs007** [comment]
/lgtm
@xiaofan-luan please take a look at

**Review by congqixia — COMMENTED** [review]
If I understand the change correctly. This implementation of merging search request cannot do failover if some shard leader failed and switch to next leader.
Also, if querynode Search/Query operation fails, there is no way to do partial success.

**Comment by czs007** [comment]
> If I understand the change correctly. This implementation of merging search request cannot do failover if some shard leader failed and switch to next leader. Also, if querynode Search/Query operation fails, there is no way to do partial success.

@congqixia  If the first attempt fails, `executeQuery` will be executed again. It will get the allocation information of the new shardLeaders, and then send the query to the corresponding shardLeader respectively. The optimization here is that if multiple shardLeaders are located in the same QueryNode, only one grpc request is required.

In addition, milvus does not seem to define the behavior of partial success at present?

**Comment by yanliang567** [comment]
image

**Comment by Zach41** [comment]
rerun ut

**Comment by Zach41** [comment]
> If I understand the change correctly. This implementation of merging search request cannot do failover if some shard leader failed and switch to next leader. Also, if querynode Search/Query operation fails, there is no way to do partial success.

I refined code to fix failover problem, please take a look again.

**Comment by Zach41** [comment]
> image

datanode memory alerting looks like caused by [Replace pulsar compression type zstd with lz4](https://github.com/milvus-io/milvus/pull/18085)

**Pull request merged**